Engine refactoring

claudiovicar opened the pull request:
@lmatiolis review, s'il te plaît.

claudiovicar committed:
Engine refactor start...

claudiovicar committed:
Extracting scenes creation and refactoring Engine calls

claudiovicar committed:
Adicionando parametro 'once' no nsn.listen

claudiovicar committed:
Refatorando Inventory e suas chamadas na Engine

claudiovicar committed:
Adding tests to the Engine

lmatiolis commented on js/Base.js:
```diff
@@ -41,21 +41,32 @@
       listeners[eventName] = [];
     }
-    listeners[eventName].push({callback: callback, scope: scope});
+    listeners[eventName].push({callback: callback, scope: scope, once: once});
```
Porque decidiu adicionar esse `once`? Só pra poluir menos retirando eventos que só são chamados uma vez ou teve algum que estava chamando mais vezes do que deveria?

claudiovicar replied:
É, pra não ficar com o listener lá que só é chamado uma vez, mesmo.

lmatiolis commented on js/Base.js:
```diff
@@ -41,21 +41,32 @@
+      listener = evtListeners[i];
+      listener.callback.call(listener.scope || this, eventObject);
+
+      if(listener.once === true){
```
Acho o `splice` confuso pra saber logo de cara o que está acontecendo... que tal criar uma function privada `_removeListenerCalledOnlyOnce` e passar esse bloco de código pra lá? Ou algo do tipo, só pra deixar mais fácil de bater o olho no código mesmo...

lmatiolis commented on js/Base.js:
```diff
@@ -41,21 +41,32 @@
+    for (var i = evtListenersCount - 1; i >= 0; i--) {
+      listener = evtListeners[i];
+      listener.callback.call(listener.scope || this, eventObject);
+
```
Acho o `splice` confuso pra saber logo de cara o que está acontecendo... que tal criar uma function privada `_removeListenerCalledOnlyOnce` e passar esse bloco de código pra lá? Ou algo do tipo, só pra deixar mais fácil de bater o olho no código mesmo...

claudiovicar replied:
Hmm, achei que tivesse ficado claro por conta do nome "once"...mas podemos fazer uma função, sim. Bom que com ela podemos aproveitar e implementar uma função "nsn.unlisten()", que remove um listener tb.

lmatiolis replied:
@claudiovicar Não ficou confuso não! ;) É só pra gente deixar mais direto mesmo o que tá rolando... fazer um `nsn.unlisten` é legal :)

lmatiolis commented on js/Engine.js:
```diff
@@ -51,33 +42,32 @@ nsn.GameEngine.prototype = {
+
+    nsn.SceneBuilder.buildScenes(this.assets["scenes.json"]);
+
+    this.setSceneAsCurrent("Apartamento");
```
Deveria ser setado em outro lugar, porque agora está hardcoded isso ai e pra alterar a currentScene inicial teríamos que alterar esse método...

claudiovicar replied:
É, tava em duvida nisso tb....penso em colocar esse evento no proprio SceneBuilder.

lmatiolis commented on js/Engine.js:
```diff
@@ -51,33 +42,32 @@ nsn.GameEngine.prototype = {
-      return this.characters[config.name];
-    }
+    var config = this.assets['characters.json'][name],
+      character;
```
Deixar identação com mesmo espaçamento do var:

```
    var config = this.assets['characters.json'][name],
        character;
```

lmatiolis commented on js/TextManager.js:
```diff
@@ -262,10 +262,10 @@ nsn.TextManager.prototype = {
       // TODO Shouldn't make direct calls to inventory
-      if(nsn.Engine.inventory.itemSelected &&
-         nsn.Engine.inventory.itemIsNotTheSameOfInventory(params.objectName)){
+      if(nsn.Inventory.itemSelected &&
```
Ainda não gosto do Inventory global, mas enquanto não arrumamos a arquitetura geral do projeto, não tem problema fazermos essa alteração...

lmatiolis commented on js/main.js:
```diff
@@ -7,8 +7,13 @@
   function initGame(){
+
     nsn.Engine = new nsn.GameEngine();
     nsn.Engine.stage = new nsn.Stage();
```
Gosto de retirarmos as criações dos objetos de dentro da Engine e colocar aqui! ;) Ainda não gosto de expormos eles assim, mas vamos resolvendo as coisas aos poucos.

lmatiolis commented on test/unit/engine.test.html:
```diff
@@ -53,10 +53,12 @@
     <!-- Load project dependencies -->
     <script type="text/javascript" src="../../build/js/lib/rsvp.min.js"></script>
     <script type="text/javascript" src="../../build/js/lib/easeljs-0.8.0.combined.js"></script>
```
Podemos já deletar caso não seja necessário mesmo! ;)

lmatiolis commented:
@claudiovicar Examen fait! :P

claudiovicar committed:
Fixing issues comments

claudiovicar committed:
Moving the creation of the default scene to the SceneBuilder

lmatiolis committed:
Fix game initialization.

This is just a quick fix to put the project working again... We
need to decouple this.

lmatiolis committed:
Fix Inventory integration spec after changes.
